File: tests/test_logger.py
"""Tests for src.utils.logger — DatasetLogger and TransitionRecord."""

import os
import csv
import tempfile

import numpy as np
import pytest

from src.utils.logger import DatasetLogger, TransitionRecord


def _make_record(t=0, episode_id=0, obs_dim=4, act_dim=2, tag="none"):
    return TransitionRecord(
        t=t,
        episode_id=episode_id,
        obs=np.random.randn(obs_dim).astype(np.float32),
        action_ctrl=np.random.randn(act_dim).astype(np.float32),
        action_dist=np.random.randn(act_dim).astype(np.float32),
        action_total=np.random.randn(act_dim).astype(np.float32),
        reward=float(np.random.randn()),
        terminated=False,
        truncated=False,
        disturbance_params=np.array([0.0, 0.1], dtype=np.float32),
        disturbance_tag=tag,
        robust_weight_target=0.5,
    )


class TestTransitionRecord:
    def test_default_values(self):
        r = _make_record()
        assert r.disturbance_tag == "none"
        assert r.robust_weight_target == 0.0 or r.robust_weight_target == 0.5

    def test_fields_accessible(self):
        r = _make_record(t=5, episode_id=3, tag="adversary")
        assert r.t == 5
        assert r.episode_id == 3
        assert r.disturbance_tag == "adversary"


class TestDatasetLogger:
    @pytest.fixture
    def logger(self, tmp_path):
        return DatasetLogger(str(tmp_path / "dataset"))

    def test_creates_log_dir(self, logger):
        assert os.path.isdir(logger.log_dir)

    def test_end_episode_empty_buffer_noop(self, logger):
        logger.end_episode()  # should not crash
        episode_dir = os.path.join(logger.log_dir, "episodes")
        assert not os.path.exists(episode_dir)

    def test_single_episode_saves_npz(self, logger):
        for t in range(10):
            logger.log(_make_record(t=t, episode_id=0))
        logger.end_episode()

        episode_dir = os.path.join(logger.log_dir, "episodes")
        npz_path = os.path.join(episode_dir, "ep_000000.npz")
        assert os.path.isfile(npz_path)

        data = np.load(npz_path, allow_pickle=True)
        assert data["obs"].shape == (10, 4)
        assert data["action_ctrl"].shape == (10, 2)
        assert data["reward"].shape == (10,)
        assert data["terminated"].shape == (10,)

    def test_csv_summary_written(self, logger):
        for t in range(5):
            logger.log(_make_record(t=t))
        logger.end_episode()

        csv_path = os.path.join(logger.log_dir, "transitions.csv")
        assert os.path.isfile(csv_path)

        with open(csv_path) as f:
            reader = csv.reader(f)
            rows = list(reader)
        assert rows[0] == ["episode_id", "length", "total_reward",
                           "has_disturbance", "mean_alpha_target"]
        assert len(rows) == 2  # header + 1 episode

    def test_multiple_episodes_increment_id(self, logger):
        for ep in range(3):
            for t in range(5):
                logger.log(_make_record(t=t, episode_id=ep))
            logger.end_episode()

        episode_dir = os.path.join(logger.log_dir, "episodes")
        assert os.path.isfile(os.path.join(episode_dir, "ep_000000.npz"))
        assert os.path.isfile(os.path.join(episode_dir, "ep_000001.npz"))
        assert os.path.isfile(os.path.join(episode_dir, "ep_000002.npz"))

    def test_csv_has_correct_disturbance_flag(self, logger):
        # Episode with disturbance
        for t in range(3):
            logger.log(_make_record(t=t, tag="adversary"))
        logger.end_episode()

        # Episode without
        for t in range(3):
            logger.log(_make_record(t=t, tag="none"))
        logger.end_episode()

        csv_path = os.path.join(logger.log_dir, "transitions.csv")
        with open(csv_path) as f:
            rows = list(csv.reader(f))
        assert rows[1][3] == "1"  # has_disturbance = True
        assert rows[2][3] == "0"  # has_disturbance = False

    def test_buffer_cleared_after_end_episode(self, logger):
        logger.log(_make_record())
        logger.end_episode()
        # buffer should be empty now — another end_episode is a no-op
        logger.end_episode()
        episode_dir = os.path.join(logger.log_dir, "episodes")
        files = os.listdir(episode_dir)
        assert len(files) == 1  # only the first episode


class TestExtractWindows:
    def test_basic_extraction(self, tmp_path):
        logger = DatasetLogger(str(tmp_path / "dataset"))
        obs_dim, act_dim, seq_len = 4, 2, 5

        # Log 2 episodes of 10 steps each
        for ep in range(2):
            for t in range(10):
                logger.log(_make_record(t=t, episode_id=ep, obs_dim=obs_dim, act_dim=act_dim))
            logger.end_episode()

        episode_dir = os.path.join(logger.log_dir, "episodes")
        obs_w, dist_l, alpha_l = DatasetLogger.extract_windows(episode_dir, seq_len)

        # 2 episodes * 10 steps = 20 windows
        assert obs_w.shape == (20, seq_len, obs_dim)
        assert dist_l.shape == (20,)
        assert alpha_l.shape == (20,)

    def test_windows_are_zero_padded(self, tmp_path):
        logger = DatasetLogger(str(tmp_path / "dataset"))
        seq_len = 5

        for t in range(3):
            logger.log(_make_record(t=t, obs_dim=4, act_dim=2))
        logger.end_episode()

        episode_dir = os.path.join(logger.log_dir, "episodes")
        obs_w, _, _ = DatasetLogger.extract_windows(episode_dir, seq_len)

        # First window: obs at t=0, padded with 4 zeros before it
        first_window = obs_w[0]
        assert first_window.shape == (seq_len, 4)
        # First 4 rows should be zero padding
        np.testing.assert_array_equal(first_window[:4], np.zeros((4, 4)))

    def test_empty_dir(self, tmp_path):
        empty_dir = str(tmp_path / "empty_episodes")
        os.makedirs(empty_dir)
        obs_w, dist_l, alpha_l = DatasetLogger.extract_windows(empty_dir, 5)
        assert obs_w.shape[0] == 0

File: cpp/tests/test_logger.cpp
// Tests for rzsm::DatasetLogger and TransitionRecord
// Mirrors Python tests/test_logger.py

#include <gtest/gtest.h>
#include <rzsm/utils/logger.hpp>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <random>
#include <vector>

namespace fs = std::filesystem;
using namespace rzsm;

static TransitionRecord make_record(int t = 0, int episode_id = 0,
                                     int obs_dim = 4, int act_dim = 2,
                                     const std::string& tag = "none") {
    std::mt19937 rng(t * 100 + episode_id);
    std::normal_distribution<float> dist(0.0f, 1.0f);

    TransitionRecord r;
    r.t = t;
    r.episode_id = episode_id;
    r.obs.resize(obs_dim);
    r.action_ctrl.resize(act_dim);
    r.action_dist.resize(act_dim);
    r.action_total.resize(act_dim);
    for (int i = 0; i < obs_dim; ++i) r.obs[i] = dist(rng);
    for (int i = 0; i < act_dim; ++i) {
        r.action_ctrl[i] = dist(rng);
        r.action_dist[i] = dist(rng);
        r.action_total[i] = dist(rng);
    }
    r.reward = dist(rng);
    r.terminated = false;
    r.truncated = false;
    r.disturbance_params = {0.0f, 0.1f};
    r.disturbance_tag = tag;
    r.robust_weight_target = 0.5f;
    return r;
}

class LoggerTest : public ::testing::Test {
protected:
    fs::path tmp_dir;
    std::unique_ptr<DatasetLogger> logger;

    void SetUp() override {
        tmp_dir = fs::temp_directory_path() / ("rzsm_test_" + std::to_string(
            std::chrono::system_clock::now().time_since_epoch().count()));
        logger = std::make_unique<DatasetLogger>(tmp_dir.string());
    }

    void TearDown() override {
        if (fs::exists(tmp_dir))
            fs::remove_all(tmp_dir);
    }
};

TEST_F(LoggerTest, CreatesLogDir) {
    EXPECT_TRUE(fs::is_directory(logger->log_dir()));
}

TEST_F(LoggerTest, EndEpisodeEmptyBufferNoop) {
    logger->end_episode();  // should not crash
    auto episode_dir = fs::path(logger->log_dir()) / "episodes";
    EXPECT_FALSE(fs::exists(episode_dir));
}

TEST_F(LoggerTest, SingleEpisodeSavesNpz) {
    for (int t = 0; t < 10; ++t)
        logger->log(make_record(t, 0));
    logger->end_episode();

    auto episode_dir = fs::path(logger->log_dir()) / "episodes";
    auto npz_path = episode_dir / "ep_000000.npz";
    EXPECT_TRUE(fs::is_regular_file(npz_path));
}

TEST_F(LoggerTest, CsvSummaryWritten) {
    for (int t = 0; t < 5; ++t)
        logger->log(make_record(t));
    logger->end_episode();

    auto csv_path = fs::path(logger->log_dir()) / "transitions.csv";
    EXPECT_TRUE(fs::is_regular_file(csv_path));

    std::ifstream f(csv_path);
    std::string line;
    int line_count = 0;
    while (std::getline(f, line)) line_count++;
    EXPECT_EQ(line_count, 2);  // header + 1 episode
}

TEST_F(LoggerTest, MultipleEpisodesIncrementId) {
    for (int ep = 0; ep < 3; ++ep) {
        for (int t = 0; t < 5; ++t)
            logger->log(make_record(t, ep));
        logger->end_episode();
    }

    auto episode_dir = fs::path(logger->log_dir()) / "episodes";
    EXPECT_TRUE(fs::is_regular_file(episode_dir / "ep_000000.npz"));
    EXPECT_TRUE(fs::is_regular_file(episode_dir / "ep_000001.npz"));
    EXPECT_TRUE(fs::is_regular_file(episode_dir / "ep_000002.npz"));
}

TEST_F(LoggerTest, CsvHasCorrectDisturbanceFlag) {
    // Episode with disturbance
    for (int t = 0; t < 3; ++t)
        logger->log(make_record(t, 0, 4, 2, "adversary"));
    logger->end_episode();

    // Episode without
    for (int t = 0; t < 3; ++t)
        logger->log(make_record(t, 1, 4, 2, "none"));
    logger->end_episode();

    auto csv_path = fs::path(logger->log_dir()) / "transitions.csv";
    std::ifstream f(csv_path);
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(f, line)) lines.push_back(line);

    ASSERT_EQ(lines.size(), 3u);  // header + 2 episodes
    // Parse CSV: has_disturbance is 4th column (index 3)
    auto get_col = [](const std::string& row, int col) {
        std::istringstream ss(row);
        std::string token;
        for (int i = 0; i <= col; ++i) std::getline(ss, token, ',');
        return token;
    };
    EXPECT_EQ(get_col(lines[1], 3), "1");  // has_disturbance = True
    EXPECT_EQ(get_col(lines[2], 3), "0");  // has_disturbance = False
}

TEST_F(LoggerTest, BufferClearedAfterEndEpisode) {
    logger->log(make_record());
    logger->end_episode();
    // Buffer should be empty now — another end_episode is a no-op
    logger->end_episode();

    auto episode_dir = fs::path(logger->log_dir()) / "episodes";
    int file_count = 0;
    for (auto& entry : fs::directory_iterator(episode_dir))
        if (entry.is_regular_file()) file_count++;
    EXPECT_EQ(file_count, 1);  // only the first episode
}

// ── TransitionRecord ────────────────────────────────────────────────

TEST(TransitionRecord, DefaultValues) {
    TransitionRecord r;
    r.disturbance_tag = "none";
    r.robust_weight_target = 0.0f;
    EXPECT_EQ(r.disturbance_tag, "none");
    EXPECT_FLOAT_EQ(r.robust_weight_target, 0.0f);
}

TEST(TransitionRecord, FieldsAccessible) {
    auto r = make_record(5, 3, 4, 2, "adversary");
    EXPECT_EQ(r.t, 5);
    EXPECT_EQ(r.episode_id, 3);
    EXPECT_EQ(r.disturbance_tag, "adversary");
}
